core-rust/state-manager/src/store/proofs_gc.rs

```diff
@@ -0,0 +1,223 @@
+        // read/write lock for relevant stages of the execution.
+        // A more robust solution should either use DB snapshots (a development direction we are
+        // considering anyway), or a more selective lock over the ledger proofs CF alone (or even
+        // over a selected *range* of the ledger proofs CF).
+
+        // Read the GC's persisted state and initialize the run:
+        let read_progress_database = self.database.read_current();
+        let to_epoch = read_progress_database
+            .max_epoch()
+            .map(|current_epoch| current_epoch.number())
+            .and_then(|current| current.checked_sub(self.most_recent_full_resolution_epoch_count))
+            .map(Epoch::of);
+        let Some(to_epoch) = to_epoch else {
+            // Nothing to GC ever, yet.
+            return;
+        };
+        let progress_started_at: LedgerProofsGcProgress = read_progress_database
+            .get_progress()
+            .unwrap_or_else(LedgerProofsGcProgress::none);
+        drop(read_progress_database);
+
+        if progress_started_at.last_pruned_epoch >= to_epoch {
+            // Nothing to GC during this run.
+            return;
+        }
+
+        info!(
+            "Starting a GC run: pruning ledger proofs up to epoch {}; current progress: {:?}",
+            to_epoch.number(),
+            progress_started_at,
+        );
+        let mut last_pruned_state_version = progress_started_at.epoch_proof_state_version;
+
+        let mut retained_proofs = 0; // only for logging purposes
+        loop {
+            let at_state_version = last_pruned_state_version
+                .next()
+                .expect("state version overflow");
+
+            // Locate the next proof that we need to retain. We use the same method and limits as
+            // the business logic responsible for outputting proven transactions:
+            let locate_proof_database = self.database.read_current();
+            let transactions_and_proof = locate_proof_database.get_txns_and_proof(
+                at_state_version,
+                MAX_TXNS_FOR_RESPONSES_SPANNING_MORE_THAN_ONE_PROOF,
+                MAX_TXN_BYTES_FOR_A_SINGLE_RESPONSE,
+            );
+            drop(locate_proof_database);
+
+            let Some((_transactions, proof)) = transactions_and_proof else {
+                error!(
+                    "A chain of transactions-without-proof from state version {} does not fit within the limits; aborting the GC",
+                    at_state_version,
+                );
+                return;
+            };
+            let header = proof.ledger_header;
+
+            // Delete all the proofs from the beginning of transactions listing up to (but
+            // excluding) the returned proof:
+            last_pruned_state_version = header.state_version;
+            let delete_database = self.database.read_current();
```

Comment:
!! Shouldn't this be `write_current()` so it contends with the ledger sync, as per your comment higher up?